--- src/main/java/net/minecraft/launchwrapper/LaunchClassLoader.java	(revision a4801b70f8a0148c6e6279ec2e91527e8019e1c8)
+++ src/main/java/net/minecraft/launchwrapper/LaunchClassLoader.java	(date 1536688577949)
@@ -54,6 +54,13 @@
         addClassLoaderExclusion("org.apache.logging.");
         addClassLoaderExclusion("net.minecraft.launchwrapper.");

+        // classloader launcher exclusions
+        addClassLoaderExclusion("launcher.");
+        addClassLoaderExclusion("com.mojang.authlib.");
+        addClassLoaderExclusion("com.mojang.util.");
+        addClassLoaderExclusion("com.google.gson.");
+        addClassLoaderExclusion("org.json.");
+
         // transformer exclusions
         addTransformerExclusion("javax.");
         addTransformerExclusion("argo.");
